Pin volume target to replicas for 1 replica volumes

tiagolobocastro opened the pull request:
fix(csi/controller): publish idempotency with 1r volumes
    
    With 1r volumes the target node may not be the same as app node. Moreover this could
    already happen if HA kicked in after publish (though probaboly not likely).
    Add frontend configuration to the volume target config, which allows us to check
    if a given node is allowed access. If it is then we can allow idempotent publish.
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    fix(node/drain): delete shutdown targets on timeout
    
    For deployer clusters, if the drain operation times out, delete the shutdown targets.
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    fix(volume/drain): during node drain don't move target away from replicas
    
    For 1-r volumes, don't move the target away from their replicas.
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    fix: pin volume target to replicas (1-r volumes)
    
    Pin the nexus together with the replica for 1-r volumes.
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    test(rust): allow specifying options and arguments
    
    Now we can run like so for example:
    ./scripts/rust/test.sh -- -Zunstable-options --report-time
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
test(rust): allow specifying options and arguments

Now we can run like so for example:
./scripts/rust/test.sh -- -Zunstable-options --report-time

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
fix: pin volume target to replicas (1-r volumes)

Pin the nexus together with the replica for 1-r volumes.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
fix(volume/drain): during node drain don't move target away from replicas

For 1-r volumes, don't move the target away from their replicas.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

chriswldenyer commented on scripts/rust/test.sh:
```diff
@@ -1,22 +1,42 @@
+      DO_ARGS="y"
+      shift;;
+    *)
+      if [ -n "$DO_ARGS" ]; then
```
IMO it's clearer as:
if [ "$DO_ARGS" == "y" ]; then

tiagolobocastro replied:
Done

chriswldenyer commented on control-plane/csi-driver/src/bin/controller/controller.rs:
```diff
@@ -467,7 +467,8 @@ impl rpc::csi::controller_server::Controller for CsiControllerSvc {
                         Ok(None)
                     },
-                    Ok(_) => Ok(Some(node_id.as_str()))
+                    Ok(_) if volume.spec.num_replicas > 1 => Ok(Some(node_id.as_str())),
```
This could do with a comment explaining what's happening here.

tiagolobocastro replied:
Done

chriswldenyer commented on control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs:
```diff
@@ -266,3 +229,111 @@ async fn check_and_drain_node(
+
+    let drain_node_id = node_spec.as_ref().id();
+    let Some(target_node) =
+        drain_target_node(context, target.node(), drain_node_id, &guarded_vol_spec)
```
(grumble) IMO I'd prefer it to be called get_drain_target_node() because at first glance the function name makes you think it is doing the drain.
But I suppose we have to stick to the rustacean way...

tiagolobocastro replied:
I agree, reading it now sounds confusing.
I'll rename to select_drain_target_node.

chriswldenyer reviewed (approved):
Looks reasonable

tiagolobocastro commented:
bors merge

tiagolobocastro commented:
bors cancel

tiagolobocastro committed:
fix(node/drain): delete shutdown targets on timeout

For deployer clusters, if the drain operation times out, delete the shutdown targets.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented:
bors merge

tiagolobocastro committed:
fix(csi/controller): publish idempotency with 1r volumes

With 1r volumes the target node may not be the same as app node. Moreover this could
already happen if HA kicked in after publish (though probaboly not likely).
Add frontend configuration to the volume target config, which allows us to check
if a given node is allowed access. If it is then we can allow idempotent publish.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented:
bors merge
